File: bfs_dfs_problems/bfs_dfs_problems__bfs_or_dfs_in_tree.cs
// 実行時間 20ms
using System.Collections.Generic;
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static int dfs = 0;
    static int y = 0;

    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        int bfs = 0;
        bool[] bb = new bool[n[0]];
        bool[] bd = new bool[n[0]];
        var g = new List<int>[n[0]];
        var q = new Queue<int>();

        y = n[2] - 1;
        bb[n[1] - 1] = true;
        foreach( int i in Range( 0, n[0] ) ) {
            g[i] = new List<int>();
        }
        foreach( int _ in Range( 0, n[0] - 1 ) ) {
            int[] a = ReadLine().Split().Select( x => int.Parse( x ) - 1 ).ToArray();

            g[a[0]].Add( a[1] );
            g[a[1]].Add( a[0] );
        }
        foreach( var l in g ) {
            l.Sort();
        }
        q.Enqueue( n[1] - 1 );
        while( 0 < q.Count ) {
            int v = q.Dequeue();

            if( v == y ) {
                break;
            } else {
                bfs++;
            }
            foreach( int i in g[v].Where( x => bb[x] != true ) ) {
                bb[i] = true;
                q.Enqueue( i );
            }
        }
        Dfs( n[1] - 1, bd, g );
        WriteLine( bfs < dfs ? "bfs" : dfs < bfs ? "dfs" : "same" );
        return;
    }

    static void Dfs( int v, bool[] s, List<int>[] g )
    {
        s[v] = true;
        if( s[y] != true ) {
            dfs++;
        }
        foreach( int i in g[v].Where( x => s[x] != true ) ) {
            Dfs( i, s, g );
        }
        return;
    }
}

File: bfs_dfs_problems/bfs_dfs_problems__bfs_or_dfs_in_tree.py
# 実行時間 10ms
from collections import deque

def dfsf(v, s, g):
    global dfs

    s[v] = True
    if s[y] != True:
        dfs += 1
    for i in g[v]:
        if s[i] != True:
            dfsf(i, s, g)

n, x, y = list(map(int, input().split()))
bfs = 0
dfs = 0
bb = [False] * n
bd = [False] * n
g = [[] for _ in [0] * n]
q = deque()

x -= 1
y -= 1
bb[x] = True
for _ in [0] * (n - 1):
    a, b = list(map(int, input().split()))

    a -= 1
    b -= 1
    g[a].append(b)
    g[b].append(a)
for l in g:
    l.sort()
q.append(x)
while 0 < len(q):
    v = q.popleft()

    if v == y:
        break
    else:
        bfs += 1
    for i in g[v]:
        if bb[i] != True:
            bb[i] = True
            q.append(i)
dfsf(x, bd, g)
print("bfs" if bfs < dfs else "dfs" if dfs < bfs else "same")
